### PR TITLE
Canonicalize MFPs below Negates

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR modifies MFP canonicalization, pushing them through Negate (and often lifting the Negate to a Union). This should come at no cost, as Negate does not modify the collection of records, and the intent is to open up better MFP reasoning by consolidating more together.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/transform/src/canonicalize_mfp.rs [hunk 1]
```diff
@@ -62,9 +62,17 @@ impl CanonicalizeMfp {
         indexes: &dyn IndexOracle,
     ) -> Result<(), crate::TransformError> {
         let mut mfp = MapFilterProject::extract_non_errors_from_expr_mut(relation);
-        relation.try_visit_mut_children(|e| self.action(e, indexes))?;
         mfp.optimize(); // Optimize MFP, e.g., perform CSE
-        Self::rebuild_mfp(mfp, relation);
+                        // Push MFPs through `Negate` operators, if encountered.
+                        // This is a first steps toward a `CanonicalizeLinear`,
+                        // which puts linear operators in a canonical representation.
```
Comment (frankmcsherry) [inline src/transform/src/canonicalize_mfp.rs:68]: Thank you `cargo fmt`. =/

File: test/sqllogictest/transform/column_knowledge.slt [hunk 2]
```diff
@@ -452,10 +452,11 @@ EXPLAIN WITH(arity, join implementations) SELECT * FROM t4 AS a1 LEFT JOIN t4 AS
...
     Union // { arity: 2 }
-      Map (123, 234) // { arity: 2 }
-        Negate // { arity: 0 }
-          Project () // { arity: 0 }
-            Get l0 // { arity: 2 }
+      Negate // { arity: 2 }
+        Project (#2, #3) // { arity: 2 }
+          Filter (#0 = 123) AND (#1 = 234) // { arity: 4 }
+            Map (123, 234) // { arity: 4 }
+              ReadStorage materialize.public.t4 // { arity: 2 }
```
Comment (frankmcsherry) [inline test/sqllogictest/transform/column_knowledge.slt:459]: This is interesting. Previously the `Map` not being pushed down meant a relation CSE opportunity was found (`l0`) and with the `Map` in the way that fails. Interesting! EQProp fixes this by removing the `Map` and `Project` entirely,

**Comment by frankmcsherry** [comment]
I have a few thoughts, @ggevay. The change isn't because `Map` being over `Negate` is harmful, as much as when canonicalizing MFPs it helps to consolidate as much M, F, and P together as possible. Also, I believe we think that a canonical representation lifts `Negate` up to a `Union` (or as close as we can manage). I think there's a different problem which is a combo of 1. `LiteralLifting` is in the logical optimization passes (despite mangling physical columns), and 2. we don't canonicalize (at least, not `CanonicalizeMFPs`) after we apply after `LiteralLifting`. I'm hesitant to block this PR on resolving that issue, which feels more like a `LiteralLifting` issue. Hopefully that makes sense!

**Comment by frankmcsherry** [comment]
The above commit looks for and removes the Map, Filter, Project pattern where we map a literal, filter by that literal, and project to keep the literal and drop the column, removing the map (and then project). It seems to clean up a fair bit, especially in `column_knowledge.slt`.

File: src/expr/src/linear.rs [hunk 3]
```diff
@@ -896,6 +896,33 @@ impl MapFilterProject {
         while prev_size.map(|p| self_size < p).unwrap_or(true) {
             // Lock in current size.
             prev_size = Some(self_size);
+
+            // We have an annoying pattern of mapping literals that already exist as columns (by filters).
+            // Try to identify this pattern, of a map that introduces an expression equated to a prior column,
+            // and then replace the mapped expression by a column reference.
```
Comment (ggevay) [inline src/expr/src/linear.rs:902]: Could you mention `LiteralLifting` in the comment, and also maybe include one of the diffs? E.g.:
```
- Project (#0, #2)
-   Filter (#1 = 1)
-     Map (1)
-       Get l0
+ Filter (#1 = 1)
+   Get l0
```

**Review by ggevay — APPROVED** [review]
LGTM!

We can resolve the things discussed in https://github.com/MaterializeInc/materialize/pull/25551#issuecomment-1967522985 in some later PR.

**Pull request merged**